### PR TITLE
Add L1/L2 authoritative UK data sources.

### PR DESCRIPTION
**Pull request by themonk911**
Only use this data from 01/04/2020 onward, the sources start to agree around then (but the match is not perfect).
UK data up to this point is from Tom White's aggregation.

### CODE REVIEW

**Comment by themonk911** [comment]
The test failure appears to be the Slovenia data source, looks unrelated to this PR.

File: src/pipelines/epidemiology/config.yaml [hunk 1]
```diff
@@ -314,6 +314,27 @@ sources:
...
+    test:
+      metadata_query: "key.str.match('GB')"
+
+  - name: pipelines.epidemiology.gb_authority.Covid19UkL2DataSource
+    fetch:
+      - url:  "https://api.coronavirus-staging.data.gov.uk/v1/data?filters=areaType=nation&structure=%7B%22areaType%22:%22areaType%22,%22areaName%22:%22areaName%22,%22areaCode%22:%22areaCode%22,%22date%22:%22date%22,%22cumCasesByPublishDate%22:%22cumCasesByPublishDate%22,%22cumDeathsByPublishDate%22:%22cumDeathsByPublishDate%22,%22cumPillarOneTestsByPublishDate%22:%22cumPillarOneTestsByPublishDate%22%7D&format=csv"
+        opts:
+          ext: csv
+        parse:
+          query: "date >= '2020-04-01'"
```
Comment (owahltinez) [inline src/pipelines/epidemiology/config.yaml:334]: For the query to have any effect, it needs to be a top-level item and it does not have to be within `parse`. I.e. make it a sibling of name and fetch.

File: src/pipelines/epidemiology/config.yaml [hunk 2]
```diff
@@ -314,6 +314,27 @@ sources:
...
       metadata_query: "key.str.match('GB_.+_.+')"
 
+  # Authoritative Data sources for GB_UK
+  - name: pipelines.epidemiology.gb_authority.Covid19UkL1DataSource
+    fetch:
+      - url: "https://api.coronavirus-staging.data.gov.uk/v1/data?filters=areaType=nation&structure=%7B%22areaType%22:%22areaType%22,%22areaName%22:%22areaName%22,%22areaCode%22:%22areaCode%22,%22date%22:%22date%22,%22cumCasesByPublishDate%22:%22cumCasesByPublishDate%22,%22cumDeathsByPublishDate%22:%22cumDeathsByPublishDate%22,%22cumPillarOneTestsByPublishDate%22:%22cumPillarOneTestsByPublishDate%22%7D&format=csv"
+        opts:
+          ext: csv
+        parse:
+          query: "date >= '2020-04-01'"
```
Comment (owahltinez) [inline src/pipelines/epidemiology/config.yaml:324]: For the query to have any effect, it needs to be a top-level item and it does not have to be within `parse`. I.e. make it a sibling of name and fetch.

File: src/pipelines/epidemiology/gb_authority.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                "date": "date",
+            },
+            drop=True,
+        )
+
+        data = data[data["subregion1_name"] != "UK"]
+        data.date = data.date.apply(lambda x: datetime_isoformat(x, "%Y-%m-%d"))
+
+        # Bug in data at time of writing.
+        data = data[data["date"] != "2020-12-07"]
```
Comment (owahltinez) [inline src/pipelines/epidemiology/gb_authority.py:46]: This is fine, although if you prefer you can also add it to the filter query: `(date > '2020-04-01') & (date != '2020-12-07')`

Comment (themonk911) [reply src/pipelines/epidemiology/gb_authority.py:46]: Yeah fair enough, I thought that it was cleaner to separate the two (one is a bug in the data, the other is an intentional cut-off point). But happy to move to the filter query.

File: src/pipelines/epidemiology/config.yaml (hunk 1 above)
Comment (themonk911) [reply src/pipelines/epidemiology/config.yaml:334]: Good catch!

**Comment by themonk911** [comment]
@owahltinez I'm reliably getting a unit test failure (both on github and locally) when trying to fetch the new source, but if I run the source through update.py on its own it collects the data fine. Is there any difference in how the unittest code collects the data?

**Comment by owahltinez** [comment]
There is no difference, the majority of the time it is an issue with the way the `metadata_query` is written. I think we should make each source declare a list of countries and each level expected in the output, that way we don't have to manually create a regex which is very error-prone.

In this case, I think the query `key.str.match('GB')` is matching all regions containing `GB` which are many. Then the subset randomly picked for testing does not contain the one you want, which is exactly `GB`. To solve this issue, you can replace the query with: `key == 'GB'`.

**Comment by themonk911** [comment]
Ah obvious in retrospect, thanks for the catch :)

File: src/pipelines/epidemiology/config.yaml [hunk 4]
```diff
@@ -314,6 +314,29 @@ sources:
...
 
+  # Authoritative Data sources for GB_UK
+  - name: pipelines.epidemiology.gb_authority.Covid19UkL1DataSource
+    fetch:
+      - url: "https://api.coronavirus-staging.data.gov.uk/v1/data?filters=areaType=nation&structure=%7B%22areaType%22:%22areaType%22,%22areaName%22:%22areaName%22,%22areaCode%22:%22areaCode%22,%22date%22:%22date%22,%22cumCasesByPublishDate%22:%22cumCasesByPublishDate%22,%22cumDeathsByPublishDate%22:%22cumDeathsByPublishDate%22,%22cumPillarOneTestsByPublishDate%22:%22cumPillarOneTestsByPublishDate%22%7D&format=csv"
+        opts:
+          ext: csv
+    parse:
+      # We only include data past 1 April 2020, prior to this there is bad agreement with our previous data source.
+      query: (date >= '2020-04-01')
```
Comment (owahltinez) [inline src/pipelines/epidemiology/config.yaml:325]: This still has no effect, `query` must be one level up (sibling of `name` and `fetch` in this case)

File: src/pipelines/epidemiology/config.yaml [hunk 5]
```diff
@@ -314,6 +314,29 @@ sources:
...
+      metadata_query: "key == 'GB'"
+
+  - name: pipelines.epidemiology.gb_authority.Covid19UkL2DataSource
+    fetch:
+      - url: "https://api.coronavirus-staging.data.gov.uk/v1/data?filters=areaType=nation&structure=%7B%22areaType%22:%22areaType%22,%22areaName%22:%22areaName%22,%22areaCode%22:%22areaCode%22,%22date%22:%22date%22,%22cumCasesByPublishDate%22:%22cumCasesByPublishDate%22,%22cumDeathsByPublishDate%22:%22cumDeathsByPublishDate%22,%22cumPillarOneTestsByPublishDate%22:%22cumPillarOneTestsByPublishDate%22%7D&format=csv"
+        opts:
+          ext: csv
+    parse:
+      # We only include data past 1 April 2020, prior to this there is bad agreement with our previous data source.
+      query: (date >= '2020-04-01')
```
Comment (owahltinez) [inline src/pipelines/epidemiology/config.yaml:336]: This still has no effect, `query` must be one level up (sibling of `name` and `fetch` in this case)

Comment (themonk911) [reply src/pipelines/epidemiology/config.yaml]: Sorry, I must've misunderstood your previous comment. Fixed now.

File: src/pipelines/epidemiology/config.yaml (hunk 4 above)
Comment (themonk911) [reply src/pipelines/epidemiology/config.yaml]: Done.

**Pull request merged**